### DISCUSSION TITLE
Multiple namespaces are created in Marquez for a simple spark job

### DISCUSSION THREAD

**Discussion by meduri30** [OpenLineage/OpenLineage#637] [Q&A] [answered]
Hello Team,
I am running OpenLineage 0.5.2 version against Spark 3.0.3 (in an EC2 instance) for testing a simple spark program that reads a local CSV file and write to a local parquet file. Marquez is hosted on a docker instance running in a different EC2 instance. And a spark scala script is getting executed over the spark-shell with below command:
`spark-shell --conf "spark.extraListeners=io.openlineage.spark.agent.OpenLineageSparkListener" \
--packages "io.openlineage:openlineage-spark:0.5.2" --conf "spark.openlineage.host=http://{some_private_ip}:5000" \
--conf "spark.openlineage.namespace=marquez_integration_test_3"`

Below is a portion of the log where the calls to marquez API are being made. It clearly shows that it is creating 2 namespaces - marquez_integration_test_3 and file. When tried to navigate to "file" namespace (which got created by default even though I have specified marquez_integration_test_3 in the command - as shown above), it is saying, "No jobs found" and when navigated to "Datasets", it is showing the name of the dataset as "complete path to the file" and when clicked on the name of the dataset, it is throwing an error - "Something went wrong".

`[29/Mar/2022:07:35:09 +0000] "GET /api/v1/namespaces/**marquez_integration_test_3**/jobs?limit=2000&offset=0 HTTP/1.1" 200 8311 "http://62.19.73.701:3000/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/99.0.4844.82 Safari/537.36" 14
[29/Mar/2022:07:35:12 +0000] "GET /api/v1/lineage/?nodeId=job:marquez_integration_test_3:spark_shell.execute_insert_into_hadoop_fs_relation_command HTTP/1.1" 200 10806 "http://62.19.73.701:3000/lineage/job/marquez_integration_test_3/spark_shell.execute_insert_into_hadoop_fs_relation_command" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/99.0.4844.82 Safari/537.36" 12
[29/Mar/2022:07:35:13 +0000] "GET /api/v1/namespaces/marquez_integration_test_3/jobs/spark_shell.execute_insert_into_hadoop_fs_relation_command/runs?limit=100&offset=0 HTTP/1.1" 200 3232 "http://62.19.73.701:3000/lineage/job/marquez_integration_test_3/spark_shell.execute_insert_into_hadoop_fs_relation_command" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/99.0.4844.82 Safari/537.36" 11
[29/Mar/2022:07:35:56 +0000] "GET /api/v1/namespaces/file/datasets?limit=2000&offset=0 HTTP/1.1" 200 3099 "http://62.19.73.701:3000/datasets" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/99.0.4844.82 Safari/537.36" 13
[29/Mar/2022:07:36:02 +0000] "GET /api/v1/lineage/?nodeId=dataset:file:%2Fhome%2Fec2-user%2Fsample_address.parquet HTTP/1.1" 200 10806 "http://62.19.73.701:3000/lineage/dataset/file/%2Fhome%2Fec2-user%2Fsample_address.parquet" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/99.0.4844.82 Safari/537.36" 16
[29/Mar/2022:07:36:11 +0000] "GET /api/v1/**namespaces/file/datasets**/home/ec2-user/sample_address.parquet/versions?limit=100&offset=0 HTTP/1.1" 404 43 "http://62.19.73.701:3000/**lineage/dataset/file**/%2Fhome%2Fec2-user%2Fsample_address.parquet" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/99.0.4844.82 Safari/537.36" 2
[29/Mar/2022:07:36:11 +0000] "GET /api/v1/lineage/?nodeId=dataset:file:%2Fhome%2Fec2-user%2Fsample_address.parquet HTTP/1.1" 200 10806 "http://62.19.73.701:3000/lineage/dataset/file/%2Fhome%2Fec2-user%2Fsample_address.parquet" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/99.0.4844.82 Safari/537.36" 15
m` 

Could some please help me with the expected behavior because the default "food_delivery" namespace is working as expected ? Also, please guide me in case I am going wrong with some configurations.
_Note:_ Only the namespaces created from OpenLineage are behaving in a different way and not working as expected.

Thank you.

Regards,
Krishna

**Comment by collado-mike** [comment]
The namespace of a dataset is unrelated to the job that reads/writes it. When you configure the namespace in your Spark configuration, you're _only_ configuring the job namespace. The Marquez UI reports no jobs under the `file` namespace, because the job ran in the `marquez_integration_test_3` namespace, as you configured.

Dataset namespaces follow the convention outlined in this naming spec - https://github.com/OpenLineage/OpenLineage/blob/main/spec/Naming.md . The namespace always reflects the source of the data - e.g., bigquery, s3, postgres, etc. In your case, it's a local file.

**Reply by meduri30** [reply]
@collado-mike Thank you very much for replying to my question :)
Got it. However, when we tried navigating to a particular dataset from lineage displayed in the `marquez_integration_test_3` namespace, Marquez was throwing an error saying that, "something went wrong" as if the respective dataset is not available in the current namespace. Could you please confirm if the behaviour is as expected from OpenLineage and needs to be taken care from Marquez, either via some configuration changes or as a code functionality (I mean fixing this at their code level) ? Or would it be the case that I might have went wrong with my installation or configuration, completely ?

**Reply by collado-mike** [reply]
It might help to provide a sample Spark script that reproduces the issue. It's possible there's a bug in the Marquez UI that's not URL encoding the file path, though from the log lines you provided, it looks right. But I can't say what's wrong without being able to look at the logs and error messages

**Reply by meduri30** [reply]
Yeah, totally makes sense. Will go thru the Marquez logs to understand more about the flow and see if they can help us with the behaviour. Will revert back with their response.
A more general question - what is the rationale behind the behaviour of keeping datasets as part of the derived namespaces and not the user specified namespace ? 
Sorry for my lack of knowledge with the documentation. Please point me to the respective location, in case this has already been answered. Thank you.

**Comment by collado-mike** [comment]
Datasets need to be named and namespaced consistently in order to maintain lineage across teams/organizations. E.g., if you have a `customer_purchases` table, it might be consumed by the Recommendations team and by the Inventory Management team. Each of those teams will probably have distinct namespaces for their jobs, but they are consuming a common dataset and that's important to show in the context of lineage. So datasets are namespaced in a consistent way regardless of the namespace of the job that is consuming/producing them.

**Reply by meduri30** [reply]
Understood. If we look from a domain-oriented lens, it makes sense to have this `customer_purchases` placed under either `customer` or `purchases` domain (in my example, the domain mapping might be poorly depicted, however the intent was more around domain segregation) and the jobs could be from a different namespace. Coming from the land of domain-driven design and Data Mesh methodologies, would like to see datasets being placed under respective domains (aka namespaces) and detach technical information such as whether a dataset is file or a big table, from the namespace conventions (I am trying not to generalise the notion but letting out my personal view).
Hence, would it be feasible to have a way to place datasets under a user-specified namespace (alongside user specified job namespace) and if not provided, could default to the current behaviour ?

**Comment by collado-mike** [comment, marked as answer]
The use case makes a lot of sense, though I tend to think this is something to be handled on the backend, rather than in the lineage collection. For example, datasets sometimes change owners (a dataset may move from `clickstream_team` to `recommendations_team` to avoid the team growing too large), ontologies may change over time (are kindle books `books` or `digital_content`?), and niche datasets may become more generally useful. These kinds of changes are difficult to manage on the lineage collection side of things (does every team that consumes the data need to update their code? how to manage namespaces for jobs that consume data from multiple domains?). I think it's easier for a dataset to have a single authoritative name, then support views that add domain logic on the backend. 

In Marquez, we have an issue that specifically calls out [renamed datasets](https://github.com/MarquezProject/marquez/issues/1512), but I think will also apply to symlinks between datasets (e.g., a Hive catalog that exposes a Postgres table or a BigQuery table that actually queries from files in GCS) and can also be used to support manually renamed and organized datasets.

**Reply by meduri30** [reply]
@collado-mike Got it. As mentioned, we might want to take it as a backend thing. Thank you very much for helping me with the information and being patient with me.
